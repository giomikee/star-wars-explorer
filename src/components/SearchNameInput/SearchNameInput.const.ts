export const SEARCH_NAME_INPUT_EVENTS = {
  nameSearched: 'name-searched',
} as const
