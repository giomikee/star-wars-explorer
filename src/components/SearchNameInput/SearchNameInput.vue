<template>
  <v-container>
    <v-skeleton-loader v-if="isLoading" type="text" />
    <v-text-field
      v-else
      v-model="searchedName"
      aria-label="Search name input"
      clearable
      label="Search name"
      persistent-clear
      @click:clear="clearSearchedName"
      @keydown.enter="$emit(SEARCH_NAME_INPUT_EVENTS.nameSearched, searchedName)"
    >
      <template #append>
        <v-btn
          aria-label="Search name"
          class="h-100"
          :disabled="searchedName.length === 0"
          prepend-icon="mdi-text-search"
          @click="$emit(SEARCH_NAME_INPUT_EVENTS.nameSearched, searchedName)"
        >
          Search
        </v-btn>
      </template>
    </v-text-field>
  </v-container>
</template>

<script lang="ts" setup>
  import { ref } from 'vue'
  import { SEARCH_NAME_INPUT_EVENTS } from './SearchNameInput.const'

  interface Props {
    isLoading: boolean
  }

  defineProps<Props>()

  const emit = defineEmits(Object.values(SEARCH_NAME_INPUT_EVENTS))

  const searchedName = ref<string>('')

  const clearSearchedName = () => {
    searchedName.value = ''

    emit(SEARCH_NAME_INPUT_EVENTS.nameSearched, '')
  }
</script>

<script lang="ts">
  export default { name: 'SearchNameInput' }
</script>
