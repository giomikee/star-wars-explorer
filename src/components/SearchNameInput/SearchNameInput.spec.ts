import { shallowMount } from '@vue/test-utils'
import { beforeAll, describe, expect, it } from 'vitest'
import { SEARCH_NAME_INPUT_EVENTS } from './SearchNameInput.const'
import SearchNameInput from './SearchNameInput.vue'

interface Component {
  searchedName: string
  clearSearchedName: () => void
}

const createWrapper = (isLoading: boolean) => shallowMount(SearchNameInput, {
  props: { isLoading },
})

describe('Given SearchNameInput component', () => {
  describe.each([true, false])('When component is rendered', isLoading => {
    describe(`And isLoading prop is ${isLoading}`, () => {
      it('Then it should match the snapshot', () => {
        const { element } = createWrapper(isLoading)

        expect(element).toMatchSnapshot()
      })
    })
  })

  describe('Given component methods', () => {
    describe('When clearSearchedName method is called', () => {
      const wrapper = createWrapper(false)
      const instance = wrapper.vm as typeof wrapper.vm & Component

      beforeAll(instance.clearSearchedName)

      it('Then searchedName should be set to empty string', () => {
        expect(instance.searchedName).toBe('')
      })

      it(`Then ${SEARCH_NAME_INPUT_EVENTS.nameSearched} event should be emitted with empty string payload`, () => {
        expect(wrapper.emitted()[SEARCH_NAME_INPUT_EVENTS.nameSearched]).toEqual([['']])
      })
    })
  })
})
