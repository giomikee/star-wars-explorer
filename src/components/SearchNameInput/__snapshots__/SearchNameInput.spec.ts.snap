// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given SearchNameInput component > When component is rendered > And isLoading prop is false > Then it should match the snapshot 1`] = `
<div
  class="v-container v-locale--is-ltr"
>
  <v-text-field-stub
    active="false"
    aria-label="Search name input"
    autofocus="false"
    clearable="true"
    clearicon="$clear"
    counter="false"
    density="default"
    direction="horizontal"
    dirty="false"
    error="false"
    errormessages=""
    flat="false"
    focused="false"
    glow="false"
    hidedetails="false"
    hidespinbuttons="false"
    iconcolor="false"
    label="Search name"
    loading="false"
    maxerrors="1"
    messages=""
    modelvalue=""
    persistentclear="true"
    persistentcounter="false"
    persistenthint="false"
    persistentplaceholder="false"
    reverse="false"
    rules=""
    singleline="false"
    tile="false"
    type="text"
    variant="filled"
  />
</div>
`;

exports[`Given SearchNameInput component > When component is rendered > And isLoading prop is true > Then it should match the snapshot 1`] = `
<div
  class="v-container v-locale--is-ltr"
>
  <v-skeleton-loader-stub
    boilerplate="false"
    loading="false"
    loadingtext="$vuetify.loading"
    type="text"
  />
</div>
`;
